### PR TITLE
CB-1778: Create SDX blueprint to handle medium scale

### PR DESCRIPTION
**Pull request by kkalvagadda**
We need to have an SDX blueprint for HA deployment to handle the medium load. Idea is to have three blueprints for SDX
1. SDX Lite Duty
2. SDX Lite Duty -HA
3. SDX Medium Duty -HA
Scope of this Jira to get an initial version and make improvements as we go.

Made sure that cloudbreak is able to set up a CM cluster using the blueprint.

### CODE REVIEW

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+        "serviceType": "KNOX",
+        "refName": "knox",
+        "roleConfigGroups": [
+          {
+            "base": true,
+            "refName": "knox-KNOX_GATEWAY-BASE",
+            "roleType": "KNOX_GATEWAY",
+            "configs": [
+              {
+                "name": "gateway_master_secret",
```
Comment (lajosrodek) [inline core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: Is this not supposed to be injected from Java code?

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+                "value": "cloudera1"
+              }
+            ]
+          },
+          {
+            "base": true,
+            "configs": [
+              {
+                "name": "idbroker_master_secret",
```
Comment (lajosrodek) [inline core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: Like above for `gateway_master_secret`, probably should come from Java.

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp (hunk 1 above)
Comment (kkalvagadda) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: Currently, it is not. "gateway_master_secret" is a mandatory config which is not injected. Will create a Jira for it. we can remove once that capability is added.

Comment (keyki) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: https://github.com/hortonworks/cloudbreak/commit/70c0b6788fe3a5d3ebe555d50c46d39848e0c4c4. It's already on master

Comment (adoroszlai) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: No need for the Jira, it was already implemented in 70c0b6788fe3a5d3ebe555d50c46d39848e0c4c4, but the version with the change may not be available yet in all environments.

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+          "hdfs-FAILOVERCONTROLLER-BASE",
+          "hdfs-JOURNALNODE-BASE",
+          "hdfs-NAMENODE-2",
+          "zookeeper-SERVER-BASE"
+        ]
+      },
+      {
+        "refName": "alpha",
+        "roleConfigGroupsRefNames": [
+          "ranger-RANGER_ADMIN-BASE",
```
[the remark this answers is not in the document]
Comment (kkalvagadda) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: There are issues adding multiple instances of ranger services. I will be creating a Jira for the same.

[the remark this answers is not in the document]
Comment (adoroszlai) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: OK, thanks for the clarification.

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+          "hdfs-DATANODE-BASE",
+          "knox-KNOX_GATEWAY-BASE",
+          "ranger-RANGER_TAGSYNC-BASE",
+          "kafka-KAFKA_BROKER-BASE",
+          "zookeeper-SERVER-BASE",
+          "hdfs-JOURNALNODE-BASE"
+        ]
+      },
+      {
+        "refName": "alpha2",
```
[the remark this answers is not in the document]
Comment (kkalvagadda) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: Like we have master, worker and compute groups for workload clusters, there was a proposal to come up group names for SDX cluster. I was told that we should be using master, alpha, and beta for now.

By default cardinality of a hostgroup is 1. I will make it explicit. Idea is to configure it as 2 but it is not supported that brought the need to come-up with alpha-1 and beta-1 group names. When cardinality is supported,  alpha1 and beta1 will be removed. Plan is to have three host groups master, alpha, beta and idbroker groups each with cardinality 2.

Please refer the sheet to an insight on what we are targetting for medium scale SD cluster.
https://docs.google.com/spreadsheets/d/105k_ilyLqQbzOA-GUxmrtbPm_ZMfJfVf13i6KR8VePQ/edit?usp=sharing

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+        "roleConfigGroupsRefNames": [
+          "kafka-GATEWAY-BASE",
+          "atlas-ATLAS_SERVER-BASE",
+          "kafka-KAFKA_BROKER-BASE",
```
[the remark this answers is not in the document]
Comment (kkalvagadda) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: We think so. In case SDX cluster kafka is used to sync information between altas, ranger and HMS. 
We have start using and see and if it good enough. If not, we can increase it.

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp (hunk 1 above)
Comment (kkalvagadda) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: created jira for this config to be injected by CB.

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp (hunk 2 above)
Comment (kkalvagadda) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: created jira for this config to be injected by CB.

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+        "serviceType": "ATLAS",
+        "roleConfigGroups": [
+          {
+            "refName": "atlas-ATLAS_SERVER-BASE",
+            "roleType": "ATLAS_SERVER",
+            "base": true,
+            "configs": [
+              {
+                "name": "atlas_kafka_bootstrap_servers",
+                "value": "{{{format-join host_groups.master format='%s:9092' }}}"
```
[the remark this answers is not in the document]
Comment (kkalvagadda) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: You are right. It has to be corrected. 
If you see a configuration in the blueprint most likely it was added because we had to add to make it work.
Let me try again and get back.

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp (hunk 1 above)
Comment (kkalvagadda) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: Sorry guys, I missed your responses and went ahead with the jira's. I will close them.

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp (hunk 6 above)
[the remark this answers is not in the document]
Comment (adoroszlai) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: Simply ssh to one of the hosts where Atlas is deployed and check `/var/log/atlas/application.log`.

[the remark this answers is not in the document]
Comment (kkalvagadda) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: ok

[the remark this answers is not in the document]
Comment (kkalvagadda) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: These configurations are needed at this point as CM is throwing NullPointerException when it is not provided and cluster setup is failing.

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp (hunk 4 above)
[the remark this answers is not in the document]
Comment (kkalvagadda) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: @adoroszlai are you fine?

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+        "serviceType": "ATLAS",
+        "roleConfigGroups": [
+          {
+            "refName": "atlas-ATLAS_SERVER-BASE",
+            "roleType": "ATLAS_SERVER",
+            "base": true,
+            "configs": [
+              {
+                "name": "atlas_kafka_bootstrap_servers",
+                "value": "{{{format-join host_groups.alpha2 format='%s:9092' }}}"
```
Comment (adoroszlai) [inline core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: I think Atlas should be pointed to both Kafka Brokers.  Pointing to local Kafka would be OK if we assume that only the entire host can go down, not the Kafka process itself, but I don't think that's true.

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp (hunk 4 above)
[the remark this answers is not in the document]
Comment (adoroszlai) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: Sure, if these names are by SDX convention, I can accept that.

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp (hunk 2 above)
Comment (adoroszlai) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: Implemented in 9450dc701d.

File: core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp (hunk 7 above)
Comment (kkalvagadda) [reply core/src/main/resources/defaults/blueprints/cdp-sdx-medium-ha.bp]: Made change to address the comment.
